src/coreclr/vm/runtimesuspension.cpp

```diff
@@ -611,35 +642,174 @@ void UnregisterTasklet(Tasklet* pTasklet)
-    while (pCurTasklet != g_pTaskletSentinel)
+    for (Tasklet* pCurStack = g_pTaskletSentinel->pTaskletNextInLiveList;
+        pCurStack != g_pTaskletSentinel;
+        pCurStack = pCurStack->pTaskletNextInLiveList)
     {
-        // Report GC pointers
-        auto pStackDataInfo = pCurTasklet->pStackDataInfo;
-        uint8_t *pLogicalRSP = pCurTasklet->pStackData - pStackDataInfo->UnrecordedDataSize;
-        uint32_t iRef;
-        for (iRef = 0; iRef < pStackDataInfo->cObjectRefs; iRef++)
+        if (pCurStack->minGeneration > condemned)
         {
-            pCallback((PTR_PTR_Object)(pLogicalRSP + pStackDataInfo->ObjectRefOffsets[iRef]), sc, 0);
-        }
-        
-        for (iRef = 0; iRef < pStackDataInfo->cByRefs; iRef++)
+            // this stack is too old to be interesting in this GC
+            continue;
+        };
+
+        Tasklet* pCurTasklet = pCurStack;
+        do
         {
-            int32_t offset = pStackDataInfo->ByRefOffsets[iRef];
-            uint32_t flags = GC_CALL_INTERIOR;
-            if (offset < 0)
+            // because of rejuvenation, some tasklets could be older than the head
+            // skip them if too old.
+            if (pCurTasklet->minGeneration > condemned)
+            {
+                // this stack is too old to be interesting in this GC
+                continue;
+            };
+
+            // Report GC pointers
+            auto pStackDataInfo = pCurTasklet->pStackDataInfo;
+            uint8_t *pLogicalRSP = pCurTasklet->pStackData - pStackDataInfo->UnrecordedDataSize;
+            uint32_t iRef;
+            for (iRef = 0; iRef < pStackDataInfo->cObjectRefs; iRef++)
             {
-                offset = -offset;
-                flags |= GC_CALL_PINNED;
+                pCallback((PTR_PTR_Object)(pLogicalRSP + pStackDataInfo->ObjectRefOffsets[iRef]), sc, 0);
             }
 
-            pCallback((PTR_PTR_Object)(pLogicalRSP + offset), sc, flags);
+            for (iRef = 0; iRef < pStackDataInfo->cByRefs; iRef++)
+            {
+                int32_t offset = pStackDataInfo->ByRefOffsets[iRef];
+                uint32_t flags = GC_CALL_INTERIOR;
+                if (offset < 0)
+                {
+                    offset = -offset;
+                    flags |= GC_CALL_PINNED;
+                }
+
+                pCallback((PTR_PTR_Object)(pLogicalRSP + offset), sc, flags);
+            }
+
+            pCurTasklet = pCurTasklet->pTaskletPrevInStack;
+        } while (pCurTasklet != NULL);
+    }
+}
+
+void AgeTasklets(int condemned, int max_gen, ScanContext* sc)
+{
+    if (!g_pConfig->TaskletAging())
+    {
+        return;
+    }
+
+    CrstHolder crstHolder(&g_taskletCrst);
+    for (Tasklet* pCurStack = g_pTaskletSentinel->pTaskletNextInLiveList; pCurStack != g_pTaskletSentinel; pCurStack = pCurStack->pTaskletNextInLiveList)
+    {
+        if (pCurStack->minGeneration > condemned)
```

Comment:
> having a pointer to a GC object which has been pinned into Gen0

I think you are referring to demotion. Sometimes GC does not want to move generation boundaries past a bunch of pinned objects with lots of empty space in between. 
If that happens GC calls `GcDemote` meaning that some objects got older, while some stayed young. In such case we could either:
1. just make min age of everything affected in this GC to be `0` for simplicity. They will re-age in later GCs.
2. dig through objects and figure the actual minimum age in each tasklet. 

I am doing `#2`, since it is easy to re-walk the objects in a tasklet, so we can do that, but no real preference between `#1` and `#2`  
([with real stacks I'd do `#1` as enumerating objects on a real stack is more involved](https://github.com/dotnet/runtime/pull/95221/files#r1409921150))

This is also the strategy used by handles - on demotion reexamine every referenced object and figure the new min age in clumps. Demotion could be somewhat expensive, but it is also rare.